Add bulk open for missing artworks and improve upload flow

SwiftExplorer567 opened the pull request:
Introduces bulk open buttons in the popup and main UI to open all missing artwork upload pages in new tabs. Enhances background.js to track and validate missing artwork URLs, and adds robust counting and auto-close logic to the upload page. Refactors API result parsing for reliability, centralizes icon and CDN URLs in constants, and improves interval cleanup and validation throughout the extension.

SwiftExplorer567 committed:
Add bulk open for missing artworks and improve upload flow

Introduces bulk open buttons in the popup and main UI to open all missing artwork upload pages in new tabs. Enhances background.js to track and validate missing artwork URLs, and adds robust counting and auto-close logic to the upload page. Refactors API result parsing for reliability, centralizes icon and CDN URLs in constants, and improves interval cleanup and validation throughout the extension.

SwiftExplorer567 committed:
Refactor missing artwork bulk upload and cleanup logs

Improves bulk upload of missing artwork by moving tab opening logic to content script and refactoring monitoring for upload completion. Removes unnecessary console logging and extensionLog calls across scripts for cleaner production code. Enhances upload counting logic for robustness and prevents double counting using per-image locks and multiple URL change detection methods.

SwiftExplorer567 committed:
Update UI feedback and remove error logging

Adds a success state to the artwork selection button by updating its class and icon. Removes console.error from the extensionError function for production use.

SwiftExplorer567 committed:
Add option to include unknown album URLs

Introduces a new setting 'Include Unknown Album URLs' to allow users to include tracks with '/_/' in their URLs when opening all missing artworks. Updates the popup UI, settings management, and missing artwork identification logic to support this feature.

SwiftExplorer567 committed:
Update README with new bulk open features

Added documentation for bulk opening missing artworks and an option to include unknown album URLs in the bulk open functionality.

neilmenon commented:
@SwiftExplorer567 woah, this looks like good stuff! Give me some time to look through these as there's significant changes here. But foremost thank you for contributing all this!

SwiftExplorer567 commented:
I Want to check algorithm for /_/ url paths for albums and tracks if there is existing album or track artwork so we can dismiss reuploads for those problematic ones. Do you have any idea. 

Before setting the extension to production please review all codebase for errors.

neilmenon commented:
@SwiftExplorer567 yeah I'll take a closer look at that. Feel free to keep pushing things up or create a separate PR. Thanks.

Yep, I'll do a re-test of everything. I played around a lot with the intervals because there were a lot of nuanced conditions and race conditions I had to deal with.

SwiftExplorer567 commented:
> @SwiftExplorer567 woah, this looks like good stuff! Give me some time to look through these as there's significant changes here. But foremost thank you for contributing all this!

Regarding the "Successfully Submitted" badge in the popup (top-right corner): previously, when multiple missing artwork submission pages were opened and submitted at once, the badge would only increase by +1 total instead of reflecting the number of actual submissions. For example, if I clicked “+” for 10 songs and then submitted them all at once using Enter + Shift+Tab, the badge would still only show +1.

I tried fixing this issue, and it seems to work now—but please double-check to confirm it’s behaving correctly in your code as well.

SwiftExplorer567 committed:
Improve missing artwork detection for unknown album URLs

Adds logic to check if artwork already exists for tracks with /_/ URLs before displaying the add artwork button. Introduces helper functions to verify artwork presence and skips elements where artwork is found, reducing unnecessary prompts.

e.g. unknown album info or malfunctioning link for artwork: https://www.last.fm/music/Golf+Trip/_/Lady+Light

It convert to https://www.last.fm/music/Golf+Trip/Lady+Light , removes the "/_" and check if there is a artwork file or not.

Did not tested fully, needs to be check in detail. But main idea is converting the link for artwork page and trying to find the artwork file.

neilmenon commented:
@SwiftExplorer567 can you let me know when you're done making changes for this feature / PR? I'm likely going to merge this into a development branch to test it and make tweaks. Don't want there to be stuff pushed after the fact. Thanks again! :)

SwiftExplorer567 commented:
Alright I'll let u know. I am trying to implement auto fix for artwork in the pages by scanning pages in order. But seems like it's a bit complicated. So it might be quit for me ill let u know in hour or two

SwiftExplorer567 commented:
<img width="465" height="600" alt="image" src="https://github.com/user-attachments/assets/eeea6ac0-1550-4a97-a521-c25b2411fa5d" />

tried this but could not handle it.

okay you can continue, i am over. @neilmenon

neilmenon commented:
@SwiftExplorer567 thanks for all your work. I’ll do my best to incorporate it into the extension. Will let you know if I have any questions.

SwiftExplorer567 commented:
> @SwiftExplorer567 thanks for all your work. I’ll do my best to incorporate it into the extension. Will let you know if I have any questions.

Enjoy developing, im here for any ideas if u feel stuck finding new ideas. Also contacted you on Instagram. Keep that in your mind. Thank you for the extension. (accidentally closed*)
